core/services/pipeline/runner.go

```diff
@@ -151,106 +192,282 @@ func (r *runner) ResultsForRun(ctx context.Context, runID int64) ([]Result, erro
 }
 
 // NOTE: This could potentially run on a different machine in the cluster than
-// the one that originally added the task runs.
-func (r *runner) processIncompleteTaskRuns() {
-	threads := int(r.config.JobPipelineParallelism())
+// the one that originally added the job run.
+func (r *runner) processUnfinishedRuns() {
+	_, err := r.processRun()
+	if err != nil {
+		logger.Errorf("Error processing unfinished run: %v", err)
+	}
+}
+
+func (r *runner) processRun() (anyRemaining bool, err error) {
+	ctx, cancel := utils.CombinedContext(r.chStop, r.config.JobPipelineMaxRunDuration())
+	defer cancel()
+
+	return r.orm.ProcessNextUnfinishedRun(ctx, r.executeRun)
+}
+
+type (
+	memoryTaskRun struct {
+		taskRun       TaskRun
+		next          *memoryTaskRun
+		nPredecessors int
+		finished      bool
+		inputs        []input
+		predMu        sync.RWMutex
+		finishMu      sync.Mutex
+	}
+
+	input struct {
+		result Result
+		index  int32
+	}
+)
+
+// results returns the results sorted by index
+// It is not thread-safe
+func (m *memoryTaskRun) results() (a []Result) {
+	inputs := make([]input, len(m.inputs))
+	copy(inputs, m.inputs)
+	sort.Slice(inputs, func(i, j int) bool {
+		return inputs[i].index < inputs[j].index
+	})
+	a = make([]Result, len(inputs))
+	for i, input := range inputs {
+		a[i] = input.result
+	}
+
+	return
+}
+
+func (r *runner) ExecuteRun(ctx context.Context, run Run) (trrs TaskRunResults, err error) {
+	return r.executeRun(ctx, r.orm.DB(), run)
+}
+
+func (r *runner) executeRun(ctx context.Context, txdb *gorm.DB, run Run) (trrs TaskRunResults, err error) {
+	logger.Debugw("Initiating tasks for pipeline run", "runID", run.ID)
+	startRun := time.Now()
+
+	if run.PipelineSpec.ID == 0 {
+		return nil, errors.Errorf("run.PipelineSpec.ID may not be 0: %#v", run)
+	}
+
+	// Find "firsts" and work forwards
+	// 1. Make map of all memory task runs keyed by task spec id
+	all := make(map[int32]*memoryTaskRun)
+	for _, tr := range run.PipelineTaskRuns {
+		if tr.PipelineTaskSpec.ID == 0 {
+			return nil, errors.Errorf("taskRun.PipelineTaskSpec.ID may not be 0: %#v", tr)
+		}
+		mtr := memoryTaskRun{
+			taskRun: tr,
+		}
+		all[tr.PipelineTaskSpec.ID] = &mtr
+	}
+
+	var graph []*memoryTaskRun
+
+	// TODO: Test with multiple and single null successor IDs
+	// https://www.pivotaltracker.com/story/show/176557536
+
+	// 2. Fill in predecessor count and next, append firsts to work graph
+	for id, mtr := range all {
```

Comment:
Any way we could abstract the similar logic in `TaskDAG#TasksInDependencyOrder()` and use it here as well?

Reply:
Probably, didnt know about that